File: lexer-ts-addon/src/lexer_addon.cpp
#include <napi.h>
#include "../../lexer/lexerpp.cpp"  // Include your lexer implementation

#include <sstream>

static Napi::Value CreateTokenArray(const Napi::CallbackInfo& info) {
	// We only need 1 argument
	if (info.Length() != 1) {
		Napi::Error::New(info.Env(), "Expected exactly one argument").ThrowAsJavaScriptException();
		return info.Env().Undefined();
	}

	// The argument should be the filename string
	if (!info[0].IsString()) {
		Napi::Error::New(info.Env(), "Expected a string argument").ThrowAsJavaScriptException();
		return info.Env().Undefined();
	}

	// Example vector: result of the lexer
	// std::vector<Token> cpparray = {
	// 	Token{token_type::comment, info[0].As<Napi::String>(), 333},
	// 	Token{token_type::identifier, "First token", 0},
	// 	Token{token_type::literal, "Second token", 100},
	// 	Token{token_type::END, "Last token", 101}
	// };

	std::istringstream content(info[0].As<Napi::String>(), std::istringstream::in | std::istringstream::binary);

	const std::vector<Token>& cpparray = lexer(content);

	// Return this c++ vector as a javasrcipt one
	Napi::Array jsarray = Napi::Array::New(info.Env(), cpparray.size());
	for (int i = 0; i < cpparray.size(); i++) {
		const Token& token = cpparray[i];
		Napi::Object jstoken = Napi::Object::New(info.Env());
		jstoken.Set("type", static_cast<int>(token.type));
		jstoken.Set("value", token.value);
		jstoken.Set("characterPosition", token.characterPosition);

		jsarray.Set(i, jstoken);
	}

	return jsarray;
}

static Napi::Object Init(Napi::Env env, Napi::Object exports) {
	exports.Set("CreateTokenArray", Napi::Function::New(env, CreateTokenArray));
	return exports;
}

NODE_API_MODULE(lexer_addon, Init);

File: lexer-ts-addon/src/lexer_addon.ts
const binding = require("bindings")("lexer_addon")

export enum token_type {
	identifier,
	keyword,
	separator,
	operator_token,
	literal,
	END,
	comment,
}

type ObjectToken = {
	type: number,
	value: string,
	characterPosition: number
}

export type Token = {
	type: token_type,
	value: string,
	characterPosition: number
};

/**
 * 
 * @param file_content The content of the file which should be parsed
 * @returns Array of tokens found in the content
 */
export const lexer = (file_content: string): Token[] => {
	const result = binding.CreateTokenArray(file_content) as ObjectToken[];

	return result.map(obj => {
        // Ensure the type is valid in the enum
        if (!(obj.type in token_type)) {
            throw new Error(`Invalid type value: ${obj.type}`);
        }

        return {
            type: obj.type as token_type, // Cast to token_type
            value: obj.value,
            characterPosition: obj.characterPosition
        };
    });
};